### PR TITLE
Range improvements

### PR DESCRIPTION
**Pull request by omninonsense**
- add O(1) `Range#size` for numeric types, with fallback to `Enumerable#size` for non-numeric ones
- add `Range#reverse_each()` and `Range::ReverseIterator`
- fixed `Range#includes?` to be *more* consistent (and thus accurate) with Ruby's behaviour
- Minor code documentation changes.

### CODE REVIEW

**Comment by omninonsense** [comment]
Pointing out that `#includes?` and `#covers?` don't behave the same in Ruby (nor should they). This is easily illustrated with this example in Ruby (whose behaviour is now matched in Crystal):

~~~ruby
("a".."z").include? "c" # => true
("a".."z").include? "cc" # => false

("a".."z").cover? "c" # => true
("a".."z").cover? "cc" # => true
~~~

File: src/range.cr [hunk 1]
```diff
@@ -187,16 +248,37 @@ struct Range(B, E)
...
+  # ("a".."zz").includes? "zz" # => true
+  # ("a"..."zz").includes? "zz" # => false
+  #```
+  def includes?(value)
+    if (@begin.is_a? Number && @end.is_a? Number) || (@begin.is_a? Char && @end.is_a? Char)
+      return covers?(value)
+    end
+
+    each {|e| return true if e == value }
+    return false
```
Comment (lbguilherme) [inline src/range.cr]: Equivalent to:

`any? {|e| e == value }`

**Comment by asterite** [comment]
Hi, thanks!

Some notes:
* add O(1) Range#size for numeric types, with fallback to Enumerable#size for non-numeric ones: this is kind of OK, although I don't know in which real scenario you will need to invoke Range#size in a real program. I'm also not sure why it has to work with floats, there's no `succ` for a Float. I find that behaviour odd in Ruby (the int optimization is ok)
* `Range#reverse_each`: OK
* fixed Range#includes? to be more consistent (and thus accurate) with Ruby's behaviour: I never liked that `include?` and `cover?` have a different meaning, it's not intuitive at all (Ruby tends to be pretty intuitive, I think here they made a mistake). For me, if a range covers some value is because it's included in the range's values, but here it's just a `<= ... <=` comparison, which you can already do with `range.begin <= ... <= range.end`, so I'm more inclined to remove the `covers?` method.

File: src/range.cr (hunk 1 above)
Comment (omninonsense) [inline src/range.cr]: Yesh, that could be changed to use `#any?`. Thanks!

**Comment by omninonsense** [comment]
Well, a Range is a (in most cases) a finite set of values,  so knowing the size might be useful (especially if there's user provided input and you don't want something crazy big). I also don't understand why Ruby does this with Floats, unless it's for use in some statistics code where you need to get the span between two float values? Really, no idea. 

I also feel like includes and covers should do the same thing, but I'm not entirely sure what the reasoning behind the current behaviour is. If a range covers an element, then in theory it should also include it. Is there a scenario where it doesn't (other than the ruby behaviour with strings).

**Comment by jhass** [comment]
Well, implementing `Comparable` and implementing `succ` are in theory orthogonal properties of an object, it can do either or both, it can also even implement them in an inconsistent manner. No, I don't have a real world example of that at hand, but I can't confidently deny it as a valid usecase.

The point is that `covers?` and `includes?` have different contracts, depending on those two properties. `covers?` is not "does the range have this value?", but "is this value bigger, smaller or equal the ends of this range?". Likewise `includes?` is not "does the range have this value?", but "is this value included in the elements yielded by calling `succ` (until `succ` returns an element equal to the end of the range)".

These are really quite distinct contracts if you think about it, thus I would consider the argument "they basically do the same just with different ways to implement it" invalid. The question is rather, do we want to provide the first contract to the caller in a convenience method, or do we want everybody to re-implement it when needed?

**Comment by asterite** [comment]
@jhass Sounds good. I'll check with @waj, though I personally never needed `Range#cover?` in Ruby, and in Crystal we can do `a <= exp <= b` pretty easily, or `a < exp < b` or `a <= exp < b`, etc., so the use cases of `Range#cover?` are pretty rare, I think.

**Comment by omninonsense** [comment]
@lbguilherme Updated the code to use `any?`. Thanks for the heads up! :)

**Comment by omninonsense** [comment]
@asterite If you already have a `Range` object, it's much more convenient/easier to use `rang.cover? x` than this, even though it may be rare:

~~~crystal
if rang.exclusive
  rang.begin <= value < rang.end
else
 rang.begin <= value <= rang.end 
end
~~~

Also, an argument for having `Range#size` would be to see the size of a result set (it could be used inside  an ORM, I suppose).

**Comment by omninonsense** [comment]
@asterite / @waj: Any updates regarding this PR, or should we abandon/close it? I am still fond of the idea of adding special handling in `#size` for numeric types, and a more Ruby-esque `#includes?` (vs it being an alias of `#covers?`).

**Comment by asterite** [comment]
@omninonsense I'm closing this. The reasons are:

1 - This PR suggests many things. Ideally a PR should suggest one thing so conversations around it are focused. Here we need to discuss 4 things. And what happens (at least to me) is that I don't like/agree some of then and then it's hard to continue with this because the PR also offers others things that I like, but there's no way to easily split stuff or accept just some changes.
2 - There's no need for O(1) Range#size, because right now it's O(1) when compiled with `--release`. 

Try compiling this with `--release --emit llvm-ir`:

```crystal
fun foo(a : Int32, b : Int32) : Int32
  (a..b).size
end

fun bar : Int32
  (10..10_000_000).size
end

foo(10, 100)
bar
```

The generated LLVM shows:

```llvm
; Function Attrs: nounwind readnone uwtable
define i32 @foo(i32 %a, i32 %b) #4 {
alloca:
  %0 = icmp sgt i32 %b, %a
  %1 = sub i32 %b, %a
  %self.0.1.val.self.0.0.val.i = select i1 %0, i32 %b, i32 %a
  %..i = select i1 %0, i32 %1, i32 0
  %2 = icmp eq i32 %self.0.1.val.self.0.0.val.i, %b
  %3 = zext i1 %2 to i32
  %4 = add i32 %3, %..i
  ret i32 %4
}

; Function Attrs: nounwind readnone uwtable
define i32 @bar() #4 {
alloca:
  ret i32 9999991
}
```

LLVM is **very smart**.

3 - Floats don't have `succ`, so I find it inconsistent that `(1.0..2.0).size` works if `(1.0..2.0).each { ... }` doesn't work. I don't think inconsistencies are good in a language. I'd also like to know a real use case for this.

4 - Since `Enumerable#includes?` basically traverses all the enumerable to find an element, and since `Range#each` traverses the enumerable in steps, using `succ`, it's equivalent to optimize this by using comparisons like `<=` and `<`. In fact I'd like to remove `#covers?` because it's an alias.

5 - Ruby's behaviour shouldn't always be copied. For example [Range#size](http://ruby-doc.org/core-2.2.3/Range.html#method-i-size) says "Returns the number of elements in the range. Both the begin and the end of the Range must be Numeric, otherwise nil is returned.". What? Only numerics? Why can't it compute it by traversing the entire range with `each`? This is inconsistent.

And I find Ruby very inconsistent in some methods, and they many times have hacks and changes like this. For example:

```crystal
p ("a".."aa").to_a
```

Ruby says: `["a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k", "l", "m", "n", "o", "p", "q", "r", "s", "t", "u", "v", "w", "x", "y", "z", "aa"]`
Crystal says: `[]`.

Ruby basically has a hardcoded rule if the range is of strings. Note that "a".succ gives "b", and `"b" > "aa"`, so iteration ends. In this case maybe Ruby's behaviour is desired, but it's not documented anywhere, I found that by looking at the source code.

6 - Range#reverse_each returning an Iterator is fine. Could you open a separate PR with just this?

Thanks!

**Comment by omninonsense** [comment]
> 2 - There's no need for O(1) Range#size, because right now it's O(1) when compiled with `--release`.
> LLVM is very smart.

It's impressive, actually! Also, I wasn't aware of the `--emit llvm-ir` flag; that's very useful. Thanks!

>3 - Floats don't have succ, so I find it inconsistent that (1.0..2.0).size works if (1.0..2.0).each { ... } doesn't work. I don't think inconsistencies are good in a language. I'd also like to know a real use case for this.
> 
> 4 - Since Enumerable#includes? basically traverses all the enumerable to find an element, and since Range#each traverses the enumerable in steps, using succ, it's equivalent to optimize this by using comparisons like <= and <. In fact I'd like to remove #covers? because it's an alias.
> 
> 5 - Ruby's behaviour shouldn't always be copied. For example Range#size says "Returns the number of elements in the range. Both the begin and the end of the Range must be Numeric, otherwise nil is returned.". What? Only numerics? Why can't it compute it by traversing the entire range with each? This is inconsistent.

> (…)

> Ruby basically has a hardcoded rule if the range is of strings. Note that "a".succ gives "b", and "b" > "aa", so iteration ends. In this case maybe Ruby's behaviour is desired, but it's not documented anywhere, I found that by looking at the source code.

Actually, I was a bit stumped by these myself. And I agree that Ruby is sometimes inconsistent.

Regarding `Range(Float, Float)#size`, I asked around a little what the idea behind it is and it seems that `(3.14..6.28).size` returns `4`; someone (in freenode#ruby) pointed out that's the number of steps between those two numbers (where step is `1.0`), and a closer look at the C implementation/logic does confirm this. Now, as to where this would be useful—beats me. I _assume_ it's to prevent constant memory reallocation, or to be able to tell if the step result will be too large before actually generating those steps?

To be fair, it is kind of confusing. I'd expect `Range(b, e)#step(n).to_a` to generate an Array of uniformly distributed steps. On the other hand, that would just add more inconsistency, probably.

**tl;dr** I agree on most points, so I'll just open a separate PR for `Range#reverse_each`. The other features aren't that important, I just thought it'd be nice to port them from Ruby, but this isn't Ruby! :+1:
